### ISSUE TITLE
Reference assemblies for C++/CLI assemblies cannot be used in WPF projects

### ISSUE THREAD

**Issue by Stefan137** [JetBrains/Refasmer#15]
I'm using Refasmer to generate reference assemblies for mixed assemblies (C++/CLI, .NET Framework 4.8) using the --public option. Building C# projects against these reference assemblies works great in general but fails when the C# project uses WPF. The build error typically reads as follows:

> C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(268,9): error MC1000: Unknown build error, 'Could not load file or assembly 'MyAssembly, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. Signature missing argument. (Exception from HRESULT: 0x801312E3)'

The error appears even for extremely simple mixed assemblies - e.g., assemblies containing just a single public type with no members. Apparently, mixed assemblies always contain an internal \<Module\> type in the global namespace. I've not yet seen the \<Module\> type in managed-only assemblies.

There is [special handling for \<Module\>](https://github.com/JetBrains/Refasmer/blob/5d39ab2dbd7367ad6936a93c9091fffb1f852146/src/Refasmer/Importer/ImportLogic.cs#L419) in the code. I've removed it for testing purposes so that \<Module\> is no longer included in the reference assemblies. These reference assemblies can then successfully be consumed in C# WPF projects. However, I do not yet know the meaning of the \<Module\> type, the impact of ignoring it, and why the special handling has been added.

**Comment by Stefan137** [comment]
Unfortunately, my proposed fix did not work under all circumstances - afterwards, some of my productive reference assemblies caused the following error in a C# WPF project:

> C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(268,9): error MC1000: Unknown build error, 'Could not load type 'System.Runtime.CompilerServices.ReferenceAssemblyAttribute' from assembly 'MyAssembly, Version=10.0.0.0, Culture=neutral, PublicKeyToken=null'.' 

I've not yet understood why .NET does not load the type from mscorlib.dll. However, this error dissapears when I use the --internals option instead of --public. In fact, reference assemblies generated by Visual Studio for C# projects contain internals as well, so this choice seems reasonable to me.

Therefore I've added special special handling for the <Module> type again, now always enforcing that <Module> is ignored. After this modification, my C# WPF projects are building successfully so far.

**Comment by Stefan137** [comment]
_Note:_ It seems that \<Module> is an artificial .NET class where global functions and variables end up[^1]. As they are not intended to be used by dependent assemblies, I guess it wouldn't hurt if \<Module> didn't show up in reference assemblies generated by Refasmer.

[^1]: https://stackoverflow.com/a/42884913/3445735

**Comment by ForNeVeR** [comment]
C++/CLI assemblies may easily depend on the global symbols from each other, so I don't think that we should just cut it out.

Could you please give some more details about your case? An MCVE would be great, but some more details is also good.

What exactly throws the error? What kind of project is it, and what's the target framework?

Perhaps also attach a msbuild binlog (collected by `dotnet build /bl`) for further investigation.

I would theorize that XAML compiler tries to load some assemblies and fails if they are refasmed, but that's a long shot.

**Comment by Stefan137** [comment]
We have a solution _A_ which contains some C++/CLI and C# library projects. And we have a solution _B_ containing C# projects only, some of which are WPF projects depending on projects from _A_. All these projects are currently targeted against .NET Framework 4.8. The projects in _B_ have not yet been updated to SDK style, so they use the old style.

Assemblies from _A_ are also used in other solutions, and we want them to be able to build AnyCPU assemblies only. Of course, we take care in the deployment step that x86 and x64 versions of the C++/CLI assemblies are available as needed at runtime. During build time, AnyCPU reference assemblies of _A_ are used. We are using Refasmer to produce the reference assemblies of the C++/CLI projects in _A_ since it is the only tool we know of for doing so.

I have created an MCVE. How should I best provide it to you?

The MCVE contains a WPF old-style project and a WPF SDK-style project for comparison reasons. For simplicity, the C# projects in the MCVE  are built for the x86 or x64 platform, not for AnyCPU. They still use a Refasmer-generated reference assembly of a C++/CLI project at build time. Interestingly, I obtain the following error (also mentioned in my initial comment) only when building the WPF old-style project, and only in the "Debug x64" configuration:

> C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(268,9): error MC1000: Unknown build error, 'Could not load file or assembly 'Mixed, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. Signature specified is zero-sized. (Exception from HRESULT: 0x80131237)'

In all other cases, the WPF project apparently builds and runs just fine.

As far as I understand, it is only the .NET API that matters in reference assemblies, not the implementation. As we are not explicitly calling anything from _&lt;Module>_, I guess we would not need it in a reference assembly - but I might be wrong.

It seens interesting to me that the presence and contents of _&lt;Module>_ in the reference assembly within the MCVE seems to depend on the configuration and platform:
- for Release x64:
![grafik](https://github.com/JetBrains/Refasmer/assets/11092727/bdcfe9f2-d00c-4337-a1d7-c2a7d5100a88)
- for Debug x64 _&lt;Module>_ is empty:
![grafik](https://github.com/JetBrains/Refasmer/assets/11092727/69c6dddc-6958-4714-ba48-bbea26924d0a)
- for Release x86 _&lt;Module>_ is missing:
![grafik](https://github.com/JetBrains/Refasmer/assets/11092727/d25022bc-521f-48cb-af96-f40df6141e9c)
- for Debug x86:
![grafik](https://github.com/JetBrains/Refasmer/assets/11092727/9ef1a3f8-3a84-428e-a9a3-d68fb4ddc4c3)

Please let me know if any further information would be helpful.

**Comment by ForNeVeR** [comment]
> I have created an MCVE. How should I best provide it to you?

Could you just drop a `.zip` archive here on GitHub? It handles small file uploads nicely (for me at least).

I'd also ask in advance for permission to use the uploaded data under the terms of Apache2 license (same as Refasmer uses), to probably push it into the repo for test purposes. If you don't allow it, that's also okay, but I'll have to reconstruct a similar example myself to use in tests.

> As far as I understand, it is only the .NET API that matters in reference assemblies, not the implementation. As we are not explicitly calling anything from <Module>, I guess we would not need it in a reference assembly - but I might be wrong.

I still disagree with this statement: C++/CLI assemblies may depend on _managed API_ of each other published as part of `<Module>`. `<Module>` is a valid way to publish managed API, even if it's hard to use from C#.

Refasmer is a tool for general purposes, and shouldn't silently discard these things.

Though I think it's okay to add an option if we've no other workaround.

From the error message, I still think we'll be able to fix it properly. It looks like Refasmer damages the native code section of a C++/CLI assembly, or something like that. It may be a bug in Cecil, but more likely a fault of our own code, and we'd need to experiment a bit to preserve the necessary semantic of `<Module>`.

**Comment by Stefan137** [comment]
Here you are: [MCVE.zip](https://github.com/JetBrains/Refasmer/files/12353243/MCVE.zip)

I hereby grant you the permission to use the content of the ZIP file under the terms of Apache License 2.0.

A few hints:
- If you don't have Refasmer on your path, you can install it as a local tool in the solution with _dotnet tool restore_.
- I've made the necessary adjustments to the C++/CLI project (called "Mixed") in PropertySheet.props where possible to make my adjustments more recognizable.
- A post-build step moves Mixed.dll into a subdirectory and executes Refasmer to produce the reference assembly in the place where Mixed.dll was before.
- The content of the C# projects mostly stems from the WPF project templates. I've added a project reference to Mixed.vcxproj and a &lt;Content> tag which copies the implementation DLL.

**Comment by Stefan137** [comment]
I agree, a more thorough solution than just ignoring _&lt;Module_ is desirable.

By the way, looking into the Debug x64 DLL that served as an input for Refasmer, I can see that _&lt;Module>_ has some public members that appeared, for example, in the Release x64 reference assembly, but not in the Debug x64 reference assembly where _&lt;Module>_ was empty:

![grafik](https://github.com/JetBrains/Refasmer/assets/11092727/04ac7d3d-8c06-4bdc-9ac7-401380ad6ecb)

**Comment by zabbius** [comment]
BTW, Refasmer does not use Cecil. Only Microsoft.CodeAnalysis.Common

**Comment by ForNeVeR** [comment]
Yep, that's my bad. Dunno why I inserted a Cecil mention there.

**Comment by ForNeVeR** [comment]
## Analysis
I was able to debug this and advanced in the investigation a little bit.

The C++/CLI assembly's `<Module>` type includes several fields with signatures like this one:
```
__m2mep@??0?$gcroot@PE$AAVString@System@@@@$$FQEAA@XZ:delegate*<gcroot<System::String ^>*, gcroot<System::String ^>*>
```

That is, a function pointer with `gcroot` as an argument.

When importing this field, we should emit `gcroot`, but our importer breaks on it: `MetadataImporter::ImportTypeHandleSignature` throws an `UnknownTypeInSignature` (which then gets ignored by the `catch` block in `MetadataImporter::ImportSignatureWithHeader`. As a result, we generate invalid assembly that doesn't even get processed by dotPeek, but I was able to extract some information using ILSpy:
<img width="837" alt="image" src="https://github.com/JetBrains/Refasmer/assets/92793/2f73967d-64c5-4934-a679-66678db70d99">

`gcroot` is an internal generic struct type defined in the same assembly, and yet the field itself is public in the (technically internal?) `<Module>` type.

## Workaround
Based on that information, I was able to produce a workaround for you: just add the `--internals` option to your Refasmer invocation. It should work (well, at least it seems to work for your MCVE).

## Conclusion
Currently, Refasmer treats the module type as public, while I'm not sure this is the right approach.

The further plan is to re-evaluate this approach: let's experiment a bit on C++/CLI assemblies to see how they expose their managed global functions to each other (if that's even a thing).

I should add that ECMA-335 (section 8.5.3) seems to underline a difference between _visible_ and _[externally-]available_ assembly member types. Which I'm not sure what to make of, so far.

**Comment by Stefan137** [comment]
Just to share what I have learned today: _\<Module>_ is the type where all global functions end up as per [ECMA-335](https://www.ecma-international.org/publications-and-standards/standards/ecma-335/) II.10.8. As per the [.NET Runtime docs](https://github.com/dotnet/runtime/blob/main/docs/design/specs/Ecma-335-Augments.md#module-initializer), _\<Module>_ is used, for example, to implement _[ModuleInitializerAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.moduleinitializerattribute)_ (available from .NET 5 upwards; see this [blog post](https://learn.microsoft.com/en-us/archive/blogs/junfeng/module-initializer-a-k-a-module-constructor) for an early dexcription of module initializers). At least, this tells us that _\<Module>_ can appear in assemblies originating from C# source code as well.

**Comment by ForNeVeR** [comment]
This is correct. But if it doesn't represent the assembly public API, then Refasmer shouldn't generally care about it, I think.
